river/TextInput.zig

```diff
@@ -0,0 +1,368 @@
+                input_method.current.delete.after_length,
+            );
+        }
+
+        text_input.wlr_text_input.sendDone();
+    }
+
+    fn handleInputMethodGrabKeyboard(
+        listener: *wl.Listener(*wlr.InputMethodV2.KeyboardGrab),
+        keyboard_grab: *wlr.InputMethodV2.KeyboardGrab,
+    ) void {
+        const self = @fieldParentPtr(Relay, "grab_keyboard", listener);
+
+        const active_keyboard = self.seat.wlr_seat.getKeyboard() orelse return;
+        keyboard_grab.setKeyboard(active_keyboard);
+        keyboard_grab.sendModifiers(&active_keyboard.modifiers);
+
+        keyboard_grab.events.destroy.add(&self.grab_keyboard_destroy);
+    }
+
+    fn handleInputMethodDestroy(
+        listener: *wl.Listener(*wlr.InputMethodV2),
+        input_method: *wlr.InputMethodV2,
+    ) void {
+        const self = @fieldParentPtr(Relay, "input_method_destroy", listener);
+
+        assert(input_method == self.input_method);
+        self.input_method = null;
+
+        const text_input = self.getFocusedTextInput() orelse return;
+        if (text_input.wlr_text_input.focused_surface) |surface| {
+            text_input.setPendingFocusedSurface(surface);
+        }
+        text_input.wlr_text_input.sendLeave();
+    }
+
+    fn handleInputMethodGrabKeyboardDestroy(
+        listener: *wl.Listener(*wlr.InputMethodV2.KeyboardGrab),
+        keyboard_grab: *wlr.InputMethodV2.KeyboardGrab,
+    ) void {
+        const self = @fieldParentPtr(Relay, "grab_keyboard_destroy", listener);
+        self.grab_keyboard_destroy.link.remove();
+
+        if (keyboard_grab.keyboard) |keyboard| {
+            keyboard_grab.input_method.seat.keyboardNotifyModifiers(&keyboard.modifiers);
+        }
+    }
+
+    pub fn getFocusableTextInput(self: *Relay) ?*TextInput {
+        var it = self.text_inputs.first;
+        return while (it) |node| : (it = node.next) {
+            const text_input = &node.data;
+            if (text_input.pending_focused_surface != null) break text_input;
+        } else null;
+    }
+
+    pub fn getFocusedTextInput(self: *Relay) ?*TextInput {
+        var it = self.text_inputs.first;
+        return while (it) |node| : (it = node.next) {
+            const text_input = &node.data;
+            if (text_input.wlr_text_input.focused_surface != null) break text_input;
+        } else null;
+    }
+
+    pub fn disableTextInput(self: *Relay, text_input: *TextInput) void {
+        const input_method = self.input_method orelse {
+            log.debug("disable text input but input method is gone", .{});
+            return;
+        };
+        input_method.sendDeactivate();
+        self.sendInputMethodState(text_input.wlr_text_input);
+    }
+
+    pub fn sendInputMethodState(self: *Relay, wlr_text_input: *wlr.TextInputV3) void {
+        const input_method = self.input_method orelse return;
+
+        // surrounding_text
+        if (wlr_text_input.active_features == 0) {
```

Comment:
Let's use the named constant here, which actually has value 0 not 1...
```suggestion
        if (wlr_text_input.active_features == wlr.TextInputV3.features.surrounding_text) {
```